### PR TITLE
tests: use pytest-bdd for acceptance

### PR DESCRIPTION
**Pull request by rikirenz**
- Here there is an example of acceptance test written with **behavioral driven development**.

### CODE REVIEW

File: tests/acceptance-bdd/conftest.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+def login(arsenic):
+    top_navigation_page.log_in('admin@inspirehep.net', '123456')
+    yield
+    top_navigation_page.log_out()
+
+
+@pytest.fixture(autouse=True, scope='module')
+def cleanup_workflows_tables(app):
+    with app.app_context():
+        obj_types = (
```
Comment (jmartinm) [inline tests/acceptance-bdd/conftest.py]: You can simplify with:

```python
WorkflowsAudit.query.delete()
WorkflowsPendingRecord.query.delete()
workflow_object_class.query().delete()
```

Comment (rikirenz) [inline tests/acceptance-bdd/conftest.py]: Could you explain why :dancing_men:

Comment (jmartinm) [inline tests/acceptance-bdd/conftest.py]: Just from the documentation - http://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.delete 

it seems like you can directly call .delete() and it will remove all rows.

http://stackoverflow.com/a/16573863/890185

File: docker-compose.test.yml [hunk 2]
```diff
@@ -83,7 +83,7 @@ services:
   acceptance:
     extends:
       service: test-service_base
-    command: py.test --driver Firefox --host selenium --port 4444 --capability browserName firefox --html=selenium-report.html tests/acceptance
+    command: py.test --driver Firefox --host selenium --port 4444 --capability browserName firefox --html=selenium-report.html tests/acceptance-bdd
```
Comment (rikirenz) [inline docker-compose.test.yml]: Only to test on travis. This commit is gonna be removed.

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    pass
+
+
+@scenario('literature_suggestion.feature', 'Accept thesis and verify the confirmation message')
+def test_verify_accepted_thesis():
+    pass
+
+
+@given('I am logged in')
+def author_user(login):
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: Now that we use BDD, can the sentence `I am logged in` become a reusable piece of code that does the login? Not to mix different styles, we can remove the fixture.

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+
+@when('I go to the holding panel list page')
+def go_to_holding_pen_literature_list():
+    Arsenic().get(os.environ['SERVER_NAME'] + '/holdingpen/list/?page=1&size=10&status=HALTED&workflow_name=HEP')
+
+
+@when(parsers.cfparse('I click on the button with id {button_id}'))
+def click_generic_button(button_id):
+    WebDriverWait(Arsenic(), 2).until(
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: Above you have a timeout of 10 and these ones here have 2. If I am correct this wait is implicit right? So it's the maximum waiting time before error.

You seem to be repeating a lot the code:

```
WebDriverWait(Arsenic(), 2).until(
EC.visibility_of_element_located((By.ID, button_id))
).click()
```

Would be good to have a helper with a default value for the wait and where you just pass the part inside the `until`

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+
+@when('I click on the accept button')
+def click_accept_button():
+    WebDriverWait(Arsenic(), 2).until(
+        EC.visibility_of_element_located((By.XPATH, '//button[@class="btn btn-warning"]'))
+    ).click()
+
+
+@when('I click on the button Skip, and fill the form manually')
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: This is more:

```
'I click on the button Skip, and uncollapse all panels'
```

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+            null,
+            XPathResult.FIRST_ORDERED_NODE_TYPE,
+            null
+        ).singleNodeValue.click()"""
+    )
+    Arsenic().show_title_bar()
+
+
+@when(parsers.cfparse('I select the value {input_select} in the select box with id {field_id}'))
+def select_generic_value(input_select, field_id):
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: This nice generic functions are now on this file `test_literature_suggestion_form.py`. Can they be moved to a file where one can find reusable `when` clauses?

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+    Select(Arsenic().find_element_by_id(field_id)).select_by_value(input_select)
+
+
+@when(parsers.cfparse('I select the value {input_select} in the subject dropdown list'))
+def select_generic_value_dropdown(input_select):
+    Arsenic().find_element_by_xpath('(//button[@type="button"])[8]').click()
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: Even though this function is called `select_generic_value_dropdown` this line is quite specific.

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+    assert globals()['_is_' + field + '_error_message_present']()
+
+
+@then(parsers.cfparse('I should not see the {field} error message'))
+def is_not_there_a_message_error(field):
+    assert not globals()['_is_' + field + '_error_message_present']()
+
+
+@when(parsers.cfparse('I insert <input> in the input box with id {field_id}'))
+@when(parsers.cfparse('I insert {input} in the input box with id {field_id}'))
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: Are both needed?

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+    assert expected_data == field.get_attribute('value')
+
+
+@then(parsers.cfparse('I should see {value_record_entry} in the record entry'))
+def check_record_detail(value_record_entry):
+    record = WebDriverWait(Arsenic(), 10).until(
+        EC.visibility_of_element_located(
+            (
+                By.XPATH,
+                '//div[@class="row hp-item ng-scope"][1]'
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: What is this selecting? (we can add an id to the html to have it more clear)

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+    assert value_record_entry in record
+
+
+@then(parsers.cfparse('I should see {value_record_detail} in the record detail'))
+def check_record_detail(value_record_detail):
+    record = WebDriverWait(Arsenic(), 10).until(
+        EC.visibility_of_element_located(
+            (
+                By.XPATH,
+                '(//div[@class="ng-scope"])[2]'
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: Same here and the selectors below, and `id` in the html will make it more readable.

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+            (
+                By.XPATH,
+                '//div[@class="row hp-item ng-scope"][1]/div/div/div[2]/holding-pen-template-handler/div[3]/a'
+            )
+        )
+    ).click()
+
+
+@then('I should see the message of confirmation')
+def click_submit_button():
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: wrong name of function

File: tests/acceptance-bdd/literature_suggestion.feature [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+    Given I am logged in
+    When I go to the literature suggestion page
+    And I insert <input> in the input box with id arxiv_id
+    Then I should see the arxiv error message
+
+  Examples: Vertical
+    | input | hep-th.9711200 |
+
+  # DOI text input
+  Scenario: Writing the correct DOI id
```
Comment (michamos) [inline tests/acceptance-bdd/literature_suggestion.feature]: "DOI" stands for Digital Object Identifier, so you can drop the "id" here and in the following.

File: tests/acceptance-bdd/literature_suggestion.feature [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+    | input | wrong | 2016-02-30 | 2016-13 |
+
+
+
+  # URL text input
+  Scenario: Writing a correct url
+    Given I am logged in
+    When I go to the literature suggestion page
+    And I click on the button Skip, and fill the form manually
+    And I insert pdf_url_correct in the input box with id url
```
Comment (michamos) [inline tests/acceptance-bdd/literature_suggestion.feature]: same question, why not do it as an example like for the other cases?

File: tests/acceptance-bdd/literature_suggestion.feature [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+  Scenario: Writing a wrong url
+    Given I am logged in
+    When I go to the literature suggestion page
+    And I click on the button Skip, and fill the form manually
+    And I insert pdf_url_wrong in the input box with id url
+    Then I should see the url error message
+
+
+  # thesis info autocomplete for supervisor institution
+  Scenario: Use the autocomplete for the supervisor institution
```
Comment (michamos) [inline tests/acceptance-bdd/literature_suggestion.feature]: Use -> Using for consistency with the previous tests?

File: tests/acceptance-bdd/literature_suggestion.feature (hunk 13 above)
Comment (rikirenz) [inline tests/acceptance-bdd/literature_suggestion.feature]: `pytest-bdd` is able to manage 2 kinds of input parameters:

- <First_type>
- {Second_type}

This is the [Gherkin](https://github.com/cucumber/cucumber/tree/master/gherkin) syntax, the variable templates are written using corner braces as _<somevalue>_ when we need to test multiple values or curly braces as _{somevalue}_ when we need to test a single value that has to be parameterized.

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+def test_import_from_doi():
+    pass
+
+
+@scenario('literature_suggestion.feature', 'Writing the wrong arXiv id')
+def test_wrong_arxiv():
+    pass
+
+
+def test_correct_arxiv():
```
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: My bad!

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 4 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: You are totally right! Actually yesterday I was reading again the selenium doc and I found this best practice:
http://selenium-python.readthedocs.io/page-objects.html?highlight=locator#locators

> One of the practices is to separate the locator strings from the place where they are being used. In this example, locators of the same page belong to same class.

File: tests/acceptance-bdd/literature_suggestion.feature [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+    And I should see 525-553 in the input box with id page_range_article_id
+    And I should see Schlegel, David J. in the input box with id authors-0-name
+    And I should see 10.1086/305772 in the input box with id doi
+    And I should see Maps of Dust Infrared Emission for Use in Estimation of Reddening and Cosmic Microwave Background Radiation Foregrounds in the input box with id title
+
+  Examples: Vertical
+    | input | 10.1086/305772 |
+
+
+  Scenario: Submit thesis and verify record detail in the list page
```
Comment (michamos) [inline tests/acceptance-bdd/literature_suggestion.feature]: This and the following tests should use realistic data, also containing unicode

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 7 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: What do you think about:  `select_value_in_literature_dropdown`

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 9 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: If we group the locator as the Selenium doc sad probably even though most of the locators are gonna use the XPATH syntax, they will have a specific name that can describe what they do. In order to make them more clear and at the same time avoid to change the code only for the tests.

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 11 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: My bad again!

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 8 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: yes, we need both!

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 6 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: you are right again! From the moment that we are going to use this framework it makes sense spend time to organize the code

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 3 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: I think this is not possible, I will try, but probably we still need to use the fixtures.

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 7 above)
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: Since this file is all about the literature suggestion form you can drop literature from the name. Something like:

`select_value_in_subject_multiselect_widget`

can work.

File: tests/acceptance-bdd/literature_suggestion.feature (hunk 16 above)
Comment (rikirenz) [inline tests/acceptance-bdd/literature_suggestion.feature]: Make it work
make it right 
make it realistic 
make it fast

I will add a test with unicode in this PR

File: tests/acceptance-bdd/literature_suggestion.feature (hunk 13 above)
Comment (kaplun) [inline tests/acceptance-bdd/literature_suggestion.feature]: But... there are no curly brackets, here.

Comment (kaplun) [inline tests/acceptance-bdd/literature_suggestion.feature]: Ah I see what you mean. The curly brackets are used then in the decorator.

Comment (rikirenz) [inline tests/acceptance-bdd/literature_suggestion.feature]: You got it!

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 3 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: @jmartinm I have found a way!! :wink:

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 8 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: We need both because with that function we are able to handle 2 different statement syntax.

File: inspirehep/bat/steps/when.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+
+@when(parsers.cfparse('I click on the button with id {button_id}'))
+def click_generic_button(button_id):
+    WebDriverWait(Arsenic(), 2).until(
+        EC.visibility_of_element_located((By.ID, button_id))
+    ).click()
+
+
+@when(parsers.cfparse('I click on the button link with text {text_value}'))
```
Comment (jmartinm) [inline inspirehep/bat/steps/when.py]: Is it just for links?  So no need to say `button`

File: tests/acceptance-bdd/literature_suggestion.feature [hunk 18]
```diff
@@ -0,0 +1,287 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2014-2017 CERN.
```
Comment (jmartinm) [inline tests/acceptance-bdd/literature_suggestion.feature]: This file is new, so just 2017

File: tests/acceptance-bdd/literature_suggestion.feature [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+    | input | 10.1086/305772 |
+
+  Scenario: Writing the wrong DOI
+    Given I am logged in
+    When I go to the literature suggestion page
+    And I insert <input> in the input box with id doi
+    Then I should see the doi error message
+
+  Examples: Vertical
+    | input | dummy:10.1086/305772 | state-doi
```
Comment (jmartinm) [inline tests/acceptance-bdd/literature_suggestion.feature]: `state-doi` is a wrong doi? We can give it a more meaningul name

File: tests/acceptance-bdd/literature_suggestion.feature [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+    And I go to the holding panel list page
+    And I should see the record in the list page
+    And I click on the button link with text Show abstract
+    Then I should see Computing in the record entry
+    And I should see My Title For Test in the record entry
+    And I should see admin@inspirehep.net in the record entry
+    And I should see Mister White; Mister Brown in the record entry
+    And I should see Lorem ipsum dolor sit amet, consetetur sadipscing elitr. in the record entry
+
+  Scenario: Submit thesis and verify record detail in the detail page
```
Comment (jmartinm) [inline tests/acceptance-bdd/literature_suggestion.feature]: `Submit thesis and verify record detail in the Holding Pen`?

File: tests/acceptance-bdd/literature_suggestion.feature [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+    And I insert Lorem ipsum dolor sit amet, consetetur sadipscing elitr. in the input box with id abstract
+    And I click on the submit button
+    And I go to the holding panel list page
+    And I should see the record in the list page
+    And I click on the button link with text My Title For Test
+    And I click on the accept button
+    Then I should see the message of confirmation
+
+
+  Scenario: Submit article and verify record detail in the list page
```
Comment (jmartinm) [inline tests/acceptance-bdd/literature_suggestion.feature]: Above you call `record detail` the detailed page of the Holding Pen, might be confusing to call it the same the entry in the search results.

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 22]
```diff
@@ -0,0 +1,287 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2014-2017 CERN.
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: Only 2017

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 23]
```diff
@@ -0,0 +1,287 @@
+# -*- coding: utf-8 -*-
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: For the name of the file to be consistent with the one above it can be `literature_suggestion.feature` and `test_literature_suggestion.py`

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+    Arsenic().get_element_with_locator(CL.SKIP_IMPORT_DATA, waiting_time=5).click()
+
+    WebDriverWait(Arsenic(), 10).until(
+        EC.text_to_be_present_in_element(
+            CL.LITERATURE_FORM,
+            'Type of Document',
+        )
+    )
+
+    Arsenic().get_element_with_locator(CL.CONFERENCE_TAB).click()
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: I would call them panel, so `CONFERENCE_PANEL` to be more accurate

File: tests/acceptance-bdd/test_literature_suggestion_form.py [hunk 25]
```diff
@@ -0,0 +1,287 @@
...
+    Arsenic().get_element_with_locator(CL.DROPDOWN_BUTTON).click()
+
+
+# when statements for holdingpen pages
+@when('I go to the holding panel list page')
+def go_to_holding_pen_literature_list():
+    Arsenic().get(os.environ['SERVER_NAME'] + '/holdingpen/list/?page=1&size=10&status=HALTED&workflow_name=HEP')
+
+
+@when('I should see the record in the list page')
```
Comment (jmartinm) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: Is this a `when` or a `then` ?

File: tests/acceptance-bdd/conftest.py [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+        yield app
+
+
+@pytest.fixture(autouse=True, scope='function')
+def arsenic(selenium, app):
+    Arsenic(selenium)
+
+
+@pytest.fixture(autouse=True, scope='function')
+def cleanup_workflows_tables(app):
```
Comment (rikirenz) [inline tests/acceptance-bdd/conftest.py]: This fixture is executed every time that we run a test in order to delete all the records. This helps to maintain all the tests atomic.

File: inspirehep/bat/locators.py [hunk 27]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class HoldingpenLiteratureListPageLocators(object):
+    RECORD_ENTRY = (By.XPATH, '//div[@class="row hp-item ng-scope"][1]')
+    SHOW_ABSTRACT_BUTTON = (By.XPATH, '//div[@class="row hp-item ng-scope"][1]/div/div/div[2]/holding-pen-template-handler/div[3]/a')
+
+
+class HoldingpenLiteratureDetailPageLocators(object):
+    PRINCIPAL_RECORD_PANEL = (By.XPATH, '(//div[@class="ng-scope"])[2]')
+    SUBMISSION_PANEL = (By.XPATH, '//p[@class="text-center ng-scope"]')
```
Comment (rikirenz) [inline inspirehep/bat/locators.py]: No they are respectively:

- This panel
image

- And this panel 
image

File: tests/acceptance-bdd/test_literature_suggestion_form.py (hunk 25 above)
Comment (rikirenz) [inline tests/acceptance-bdd/test_literature_suggestion_form.py]: By default pytest-bdd does not allow to mix the syntax of the steps, that should remain always the following:

- Given
- Then
- When

buuuuuut actually I discovered this PR: https://github.com/pytest-dev/pytest-bdd/pull/130 where they have added a way to relax the strict Gherkin syntax. This allows to mix different steps even thought they do not follow the classic order `Given when then`. 
 
What do you think? Should we allow to use a mixed syntax? 

IMHO YES! Morover in this issue https://github.com/pytest-dev/pytest-bdd/issues/118 a guy that was writing selenium test sad that was necessary for this kind of tests use a mixed syntax.

**Comment by rikirenz** [comment]
Guys I completely forget about this PR! Do you know the state of this PR? Am I missing something? Maybe there is something wrong that I missed...

**Comment by kaplun** [comment]
Hi @rikirenz! I believe one issue that arose was that formally these are not end-to-end testing in the sense that are not testing UI->Python->DB but UI->Python->DB->Python->UI. I am not sure if this is actually a blocker for this PR, though. @jacquerie might want to say more.
Another potential issue is that of course these type of tests are making many assumption on the UI (which is unavoidable IMHO) therefore making more rigid the possibility of updating it in an *agile* (:dancer:) way. But I don't see a way out of this. Tests involving the UI are going to always impose some rigidity on the UI itself.

**Comment by jacquerie** [comment]
The point of BDD is to express tests in the domain of the user. This PR instead is still using the language of the developer: users don't think in terms of "when I push this button and I fill this text", but in terms of "when I add a new INSPIRE category".

Another objective of this PR was to simplify the complicated architecture of the acceptance tests, and instead it is making it more complicated, as it's adding one extra level of abstraction (which, as a I said above is not abstracting at all, it's merely recasting the below level with other code).

Because of these two main reasons, I'm going to close this PR. Sorry!
